### PR TITLE
Implement selling

### PR DESCRIPTION
**Pull request by bmacer**

### CODE REVIEW

**Review by 007nh — COMMENTED** [review]
@bmacer this overall looks really good. Successfully compiling with no build issues as well.

Please add:

- event emitters for all events
- unit tests to cover events

Some things to keep in mind for the next upgrade to the pallet

- Potential batch purchasing(will need to work through this on the product side first)
- Ability to charge a fee for the sale. The fee collector should be able to be assigned a designated address and upon completion of a sale, a percentage of the sale tokens should be deposited to the fee collector address then the actual seller of the asset(I believe this is the immediate need after the initial implementation)

Once we have events in, I will do a final review and then we can get this merged in

File: pallets/collectibles/src/lib.rs [hunk 1]
```diff
@@ -277,6 +289,22 @@ pub mod pallet {
...
+		/// Collectible sale price mismatch
+		CollectibleSalePriceMismatch,
+		/// Purchaser already owns collectible
+		YouAlreadyOwnThisCollectible,
+		/// Balance transfer failed during collectible purchase
+		CollectibleBalanceTransferFailed,
+		/// Transfer failed during collectible purchase
+		CollectibleTransferFailed,
+		/// Not enough funds for collectible purchase
+		NotEnoughFundsForPurchase,
```
[the remark this answers is not in the document]
Comment (bmacer) [reply pallets/collectibles/src/lib.rs:307]: The more error messages the merrier, in my opinion.  Though I'd always want as specific as possible--where would you want to place this general error message?

File: pallets/collectibles/src/lib.rs [hunk 2]
```diff
@@ -265,6 +265,24 @@ pub mod pallet {
 		/// Attribute metadata has been cleared for an asset class or instance.
 		/// \[ class, maybe_instance, key, maybe_value \]
 		AttributeCleared(T::ClassId, Option<T::InstanceId>, BoundedVec<u8, T::KeyLimit>),
+		/// Collectible listed for sale
+		/// \[ seller, class, instance, price \]
+		CollectibleListedForSale(
+			T::AccountId, 
+			T::ClassId, 
+			T::InstanceId, 
+			<<T as pallet::Config<I>>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance),
```
Comment (bmacer) [inline pallets/collectibles/src/lib.rs:274]: One issue I found was that I had to use this ugly representation of Balance numerous times.  I couldn't figure out the proper way to alias this once and then refer to that alias throughout the code.

**Comment by bmacer** [comment]
i did notice that while we restrict "price" to positive, "0" is considered positive and isn't prevented.  as such, a collectible can be listed for a price of 0.  i didn't implement this change because i'm not sure which way we want to go with this.  perhaps we *do* want to allow zero-priced sales.

**Pull request merged**